mycore-base/src/main/java/org/mycore/frontend/jersey/filter/MCRSessionHookFilter.java

```diff
@@ -47,20 +48,68 @@
     @Context
     private HttpServletResponse httpResponse;
 
+    private final static String ATTR = MCRSessionHookFilter.class.getName() + ".session";
+
     private static final Logger LOGGER = LogManager.getLogger(MCRSessionHookFilter.class);
 
     @Override
-    public void filter(ContainerRequestContext request) throws IOException {
+    public void filter(ContainerRequestContext request) {
+        MCRSessionMgr.unlock();
         MCRSession session = MCRServlet.getSession(httpRequest);
+        request.setProperty(ATTR, session);
         MCRSessionMgr.setCurrentSession(session);
         LOGGER.info(MessageFormat.format("{0} ip={1} mcr={2} user={3}", request.getUriInfo().getPath(),
             MCRFrontendUtil.getRemoteAddr(httpRequest), session.getID(), session.getUserInformation().getUserID()));
         MCRFrontendUtil.configureSession(session, httpRequest, httpResponse);
     }
 
     @Override
-    public void filter(ContainerRequestContext request, ContainerResponseContext response) throws IOException {
-        MCRSessionMgr.releaseCurrentSession();
+    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
+        MCRSessionMgr.unlock();
+        MCRSession requestSession = (MCRSession) requestContext.getProperty(ATTR);
+        if (responseContext.hasEntity()) {
+            responseContext.setEntityStream(new ProxyOutputStream(responseContext.getEntityStream()) {
+                @Override
+                public void close() throws IOException {
+                    LOGGER.debug("Closing EntityStream");
+                    try {
+                        super.close();
+                    } finally {
+                        releaseSessionIfNeeded(requestSession);
+                        LOGGER.debug("Closing EntityStream done");
+                    }
+                }
+            });
+        } else {
+            LOGGER.debug("No Entity in response, closing MCRSession");
+            releaseSessionIfNeeded(requestSession);
+        }
+    }
+
+    private static void releaseSessionIfNeeded(MCRSession requestSession) {
+        if (MCRSessionMgr.hasCurrentSession()) {
+            MCRSession currentSession = MCRSessionMgr.getCurrentSession();
+            try {
+                if (currentSession.isTransactionActive()) {
```

Comment:
shouldn't this be part of MCRSessionMgr.releaseCurrentSession?

Reply:
It could be implemented that way that transaction are automatically finished when a session is released (or closed). This should get another ticket and would be part of other code changes. That would definitely make the code simply to read by hiding boilerplate code. We have to to think carefully of any side-effects this may have, but I think there are none.

But at current state this code is handling corner cases better than before and cannot be written simpler - suggestions are welcome.

Reply:
It _**could**_ be part of releaseCurrentSession() and would simplify this code a lot. This change is out of scope of this ticket and should be an extra one as this would include changes in other parts. We should carefully think about side-effects though. Currently, I think there are none. But definitely a good point we should keep in mind.